runtime/tracing/tracing-core/common/src/aws/smithy/kotlin/runtime/tracing/TraceSpanExt.kt

```diff
@@ -0,0 +1,219 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package aws.smithy.kotlin.runtime.tracing
+
+import aws.smithy.kotlin.runtime.io.use
+import aws.smithy.kotlin.runtime.logging.Logger
+import aws.smithy.kotlin.runtime.time.Instant
+
+/**
+ * Creates a child span of this [TraceSpan] and passes it to the given [block]. The span is closed when the block
+ * completes (regardless of whether the block terminates normally or exceptionally).
+ * @param id The id for the new span. IDs should be unique among sibling spans within the same parent.
+ * @param block The block to execute with the new child span passed as an argument.
+ */
+public inline fun <T> TraceSpan.withChildSpan(id: String, block: (TraceSpan) -> T): T = child(id).use { block(it) }
+
+/**
+ * Records a single new event that has occurred within the logical context of this span.
+ * @param event The event to record.
+ */
+public fun TraceSpan.postEvent(event: TraceEvent): Unit = postEvents(listOf(event))
+
+/**
+ * Logs a message in this [TraceSpan].
+ * @param level The level (or severity) of this event
+ * @param sourceComponent The name of the component that generated the event
+ * @param ex An optional exception which explains the message
+ * @param content A lambda which provides the content of the message. This content does not need to include any data
+ * from the exception (if any), which may be concatenated later based on probe behavior.
+ */
+public fun TraceSpan.log(level: EventLevel, sourceComponent: String, ex: Throwable? = null, content: () -> Any?) {
+    val event = TraceEvent(
+        level,
+        sourceComponent,
+        Instant.now(),
+        "thread-id", // TODO
```

Comment:
**question:** what needs to be done here? is it possible to fetch the thread ID at this point, or will it come in a future PR?

Reply:
That's a great question and I don't yet have a great answer. We have a reference architecture which specifies that "thread ID" should be part of trace events but no concrete plan to get it (or use it) yet. [Aaron mentioned above](https://github.com/awslabs/smithy-kotlin/pull/737#discussion_r1025629428) that we might consider putting the coroutine name in this field but I don't yet know how we'd get that in practice.